domain/permission/state/state.go

```diff
@@ -195,25 +196,153 @@ func (s *State) UpsertPermission(ctx context.Context, args permission.UpsertPerm
-	return corepermission.UserAccess{}, errors.NotImplementedf("ReadUserAccessForTarget")
+	var userAccess corepermission.UserAccess
+	db, err := s.DB()
+	if err != nil {
+		return userAccess, errors.Trace(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		user, err := findUserByName(ctx, tx, subject)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		userAccess = user.toCoreUserAccess()
+
+		// Based on the grant to and grant from, find the permission,
+		// then the access type of it.
+		accessType, permissionUUID, err := findAccessType(ctx, tx, target.Key, user.UUID)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		userAccess.Access = corepermission.Access(accessType)
+		userAccess.PermissionID = permissionUUID
+		userAccess.Object = objectTag(target)
+		return nil
+	})
+	if err == nil {
+		return userAccess, nil
+	}
+	return corepermission.UserAccess{}, errors.Trace(domain.CoerceError(err))
 }
 
 // ReadUserAccessLevelForTarget returns the subject's (user) access level
 // for the given user on the given target.
 func (s *State) ReadUserAccessLevelForTarget(ctx context.Context, subject string, target corepermission.ID) (corepermission.Access, error) {
-	return corepermission.NoAccess, errors.NotImplementedf("ReadUserAccessLevelForTarget")
+	userAccessType := corepermission.NoAccess
+	db, err := s.DB()
+	if err != nil {
+		return userAccessType, errors.Trace(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		userUUID, err := userUUIDForName(ctx, tx, subject)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		// Based on the grant to and grant from, find the permission,
+		// then the access type of it.
+		accessType, _, err := findAccessType(ctx, tx, target.Key, userUUID)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		userAccessType = corepermission.Access(accessType)
+		return nil
+	})
+	if err == nil {
+		return userAccessType, nil
+	}
+	return userAccessType, errors.Trace(domain.CoerceError(err))
 }
 
 // ReadAllUserAccessForUser returns a slice of the user access the given
 // subject's (user) has for any access type.
 func (s *State) ReadAllUserAccessForUser(ctx context.Context, subject string) ([]corepermission.UserAccess, error) {
-	return nil, errors.NotImplementedf("ReadAllUserAccessForUser")
+	db, err := s.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	var (
+		permissions []readUserPermission
+		user        user
+	)
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		user, err = findUserByName(ctx, tx, subject)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		permissions, err = readUsersPermissions(ctx, tx, user.UUID)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		permissions, err = grantOnType(ctx, tx, permissions)
+		if err != nil {
+			return errors.Trace(err)
+		}
```

Comment:
Let's fill in the outer value when we have all the data.

```suggestion
		userPermissions, err := readUsersPermissions(ctx, tx, user.UUID)
		if err != nil {
			return errors.Trace(err)
		}

		permissions, err = grantOnType(ctx, tx, userPermissions)
		if err != nil {
			return errors.Trace(err)
		}
```